### DISCUSSION TITLE
CSharp grammar never sets permissibleModifiers - is that correct?

### DISCUSSION THREAD

**Discussion by vsajip** [javacc21/javacc21#208] [Q&A] [answered]
https://github.com/javacc21/javacc21/blob/933ce5dcc3f8caf82aa63d7e05f144394a867c19/examples/csharp/CSharp.javacc#L209-L212
Unlike in the Java grammar,
https://github.com/javacc21/javacc21/blob/933ce5dcc3f8caf82aa63d7e05f144394a867c19/examples/java/Java.javacc#L199-L201
Nothing in the C# grammar ever sets `permissibleModifiers` - and the actions that reference it are useless since they do a check for `null` and so always effectively do nothing. Example:
https://github.com/javacc21/javacc21/blob/933ce5dcc3f8caf82aa63d7e05f144394a867c19/examples/csharp/CSharp.javacc#L232-L234
So, is our C# parsing too forgiving of incorrect modifiers (which you might not see in typical source files which actually compile)?

**Comment by revusky** [comment, marked as answer]
Well, I think what happened is that I made a mental note to get the whole modifiers thing in a proper state at a later point, but then I must have forgotten about it. That sometimes happens. I figured I'd get the C sharp parser parsing everything and only after that go for maximal correctness, nailing down the permissibleModifiers stuff, but then I was so tired after getting the grammar to parse everything I could throw at it, that the `permissibleModifiers` issue just fell through the cracks, so the existing grammar just accepts any sequence of modfiiers.

>Nothing in the C# grammar ever sets permissibleModifiers

Well, yeah... if you never set permissibleModifiers, it just accepts any sequence of modifiers, I guess, so, for one thing, it will accept combinations that make no sense, like `public private` or `abstract final` or whatever.  There's another nuance in all this, which is the whole question of the lookahead phase vs. actual parsing. Quite possibly a practical approach is frequently to be pretty lax in a lookahead phase, maybe just accepting any sequence of modifiers and then getting strict at the parsing stage.

So, I mean, you have:

       public private void someMethod() {....}

That's obviously erroneous, but at the lookahead phase, we let that pass, and it identifies that we should enter the MethodDeclaration production, but then when it does enter that production to parse it, it does set `permissibleModifiers` and then it chokes on the invalid combination -- inside the Modifiers production.

I was just looking again at how the Java grammar handles these sorts of things and I'm not sure how consistent it is. It might do to have a look at it. (And then turn to the CSharp grammar and see how that should work.) 

And, you see in the `MethodDeclaration` produciton `permissibleModifiers` is only being set for parsing, but not for lookahead. If it was set for lookahead, the code block would have a `#` at the end. So the approach taken is to accept any sequence of modifiers at the lookahead stage, i.e. just to figure out whether we are going to parse (or try to) a MethodDeclaration, but then when we enter the production to parse, the `permissibleModifiers =...` block is actually executed, so `permissibleModifiers` gets set and things like `private public` and so on  actually get caught at that point.

As you doubtless realize, these are subtleties that are about impossible to express in legacy JavaCC. I wonder how ANTLR handles this sort of thing. My suspicion is that we're way ahead of them.

Well, I guess the point is that if you want to get the C\# grammar more correct, by all means! Like I said, I surely meant to do this but then didn't get around to it.

>So, is our C# parsing too forgiving of incorrect modifiers (which you might not see in typical source files which actually compile)?

Well, it pretty much certainly is. Basically, I was obsessed with the issue of accepting correct code. The other problem of rejecting incorrect code, I figured I'd turn to that later, but never did. And now you bring this up, and the whole issue had slipped my mind!
